### PR TITLE
feat: context managers in ServiceBrowser and AsyncServiceBrowser

### PR DESCRIPTION
**Pull request by Tenebrosus3141**
I started using this package for the first time recently, and when trying to use the [async browser example](https://github.com/python-zeroconf/python-zeroconf/blob/master/examples/async_browser.py), I found the requirement to use `AsyncServiceBrowser.async_cancel()` at the end of everything. After discovering that AsyncZeroconf has an async context manager, I found that AsyncServiceBrowser did not have its own.

I tried my best to replicate #284 (#177), however I am not sure how to go about adding test coverage. Apologies for not adding any.

Here's an example of how my code is simplified using the context manager (in Python 3.11):
```
async def main():
    services = ["_example._tcp.local."]
    async with (
        AsyncZeroconf() as aiozeroconf,
        AsyncServiceBrowser(aiozeroconf.zeroconf, services, handlers=[
            lambda **kwargs: tg.create_task(async_display_service_info(**kwargs)) and None
        ]),
        asyncio.TaskGroup() as tg,
    ):
        await asyncio.sleep(5)


if __name__ == '__main__':
    asyncio.run(main())
```

### CODE REVIEW

**Comment by bdraco** [comment]
Can you add a test for these?

Also generally its not recommended to have short running instances if you can avoid it since it churns the network traffic quite a bit. There are obviously uses cases for it, and people do get it wrong a lot (and leak browsers) so I'm inclined to move this forward

**Comment by Tenebrosus3141** [comment]
> Can you add a test for these?

I added tests now

Though there seems to be a race condition between calling `ServiceBrowser.cancel()` or `AsyncServiceBrowser.async_cancel()`  and checking the value of `_ServiceBrowserBase.done`, but I believe that may be out of scope for this PR

> Also generally its not recommended to have short running instances if you can avoid it since it churns the network traffic quite a bit. There are obviously uses cases for it, and people do get it wrong a lot (and leak browsers) so I'm inclined to move this forward

My use case may be for that, my own project intends to just run the browser briefly at startup

**Comment by bdraco** [comment]
Async_cancel should not race

Cancel should but you can do asyncio.sleep(0) to ensure the call soon is run

**Comment by Tenebrosus3141** [comment]
I put the call to `asyncio.sleep(0)` in the synchronous browser test and removed it from the async one now
It should be ready now

File: tests/services/test_browser.py [hunk 1]
```diff
@@ -70,6 +71,36 @@ class MyServiceListener(r.ServiceListener):
     zc.close()
 
 
+# This test uses asyncio because ServiceBrowser.cancel uses call_soon_threadsafe
+@pytest.mark.asyncio
+async def test_service_browser_cancel_context_manager():
```
Comment (bdraco) [inline tests/services/test_browser.py:76]: I think you could leave this being a sync test

File: tests/services/test_browser.py [hunk 2]
```diff
@@ -70,6 +71,36 @@ class MyServiceListener(r.ServiceListener):
...
+
+    browser = r.ServiceBrowser(zc, type_, None, listener)
+
+    assert cast(bool, browser.done) is False
+
+    with browser:
+        pass
+
+    # ensure call_soon_threadsafe is run
+    await asyncio.sleep(0)
```
Comment (bdraco) [inline tests/services/test_browser.py:97]: Maybe use `asyncio.run_coroutine_threadsafe(asyncio.sleep(0), zc.loop).result()` instead to keep the test sync so the sync versions are used in the expected way

**Comment by Tenebrosus3141** [comment]
> Maybe use `asyncio.run_coroutine_threadsafe(asyncio.sleep(0), zc.loop).result()` instead to keep the test sync so the sync versions are used in the expected way

That makes more sense, I wasn't sure how to do that until you pointed out how, thanks

**Review by bdraco — APPROVED** [review]
Thanks @Tenebrosus3141

**Pull request merged**